### PR TITLE
Fix amp-form server side rendering

### PR DESCRIPTION
**Pull request by alabiaga**
b/123310281

### CODE REVIEW

**Comment by alabiaga** [comment]
cc\ @gotcwang

**Comment by happymark01** [comment]
Hi

**Comment by alabiaga** [comment]
@choumx note that tests are pending. But as discussed offline please do take a look to see if there is an API change to suggest, as I expressed my concern on maintenance when SSR is extended to other components.

File: extensions/amp-form/0.1/amp-form.js [hunk 1]
```diff
@@ -1036,27 +1036,34 @@ export class AmpForm {
...
-                    container.appendChild(rendered);
-                    const renderedEvent = createCustomEvent(
-                        this.win_,
-                        AmpEvents.DOM_UPDATE,
-                        /* detail */ null,
-                        {bubbles: true});
-                    container.dispatchEvent(renderedEvent);
-                  });
-            });
+        if (this.ssrTemplateHelper_.isSupported()) {
```
Comment (choumx) [inline extensions/amp-form/0.1/amp-form.js]: Looks like we can move this conditional into a new function for ssr-template-helper.js.

File: extensions/amp-form/0.1/amp-form.js [hunk 2]
```diff
@@ -819,7 +819,7 @@ export class AmpForm {
     // Construct the fetch response to reuse the methods in-place for
     // amp CORs validation.
     this.ssrTemplateHelper_.verifySsrResponse(this.win_, response, request);
-    return this.handleSubmitSuccess_(tryResolve(() => response['html']));
+    return this.handleSubmitSuccess_(tryResolve(() => response));
```
[the remark this answers is not in the document]
Comment (alabiaga) [reply extensions/amp-form/0.1/amp-form.js]: handleSubmitSuccess expects a JsonObject, so here I just pass in the JSON object and just extract the html of type string when rendering the template.

File: src/ssr-template-helper.js [hunk 3]
```diff
@@ -94,6 +94,24 @@ export class SsrTemplateHelper {
         ));
   }
 
+  /**
+   * @param {!Element} container
+   * @param {!JsonObject} data
+   * @return {!Promise}
+   */
+  renderTemplate(container, data) {
```
Comment (choumx) [inline src/ssr-template-helper.js]: We should also use this in amp-list, right? Sharing more code was the point here.

File: extensions/amp-form/0.1/test/test-amp-form.js [hunk 4]
```diff
@@ -279,6 +279,64 @@ describes.repeated('', {
...
+                  Promise.resolve({
+                    data: '<div>much success</div>',
+                  }));
+          const renderedTemplate = createElement('div');
+          renderedTemplate.innerText = 'much success';
+          sandbox.stub(ampForm.ssrTemplateHelper_.templates_, 'findTemplate')
+              .returns(template);
+          sandbox.stub(ampForm.templates_, 'findAndSetHtmlForTemplate')
+              .returns(Promise.resolve({html: renderedTemplate}));
+          sandbox.stub(ampForm.templates_, 'findAndRenderTemplate');
```
Comment (choumx) [inline extensions/amp-form/0.1/test/test-amp-form.js]: We shouldn't be testing behavior of another class (`SsrTemplateHelper`). 

We should instead verify that the new function is called, and then add a unit test in test-ssr-template-helper.js.

File: extensions/amp-form/0.1/amp-form.js [hunk 5]
```diff
@@ -810,9 +810,9 @@ export class AmpForm {
 
   /**
    * Transition the form to the submit success state.
-   * @param {!JsonObject|string|undefined} response
+   * @param {!JsonObject} response
    * @param {!FetchRequestDef} request
-   * @return {!Promise}
+   * @return {!Promise<string>}
```
Comment (choumx) [inline extensions/amp-form/0.1/amp-form.js]: This type is incorrect. Doesn't `gulp check-types` complain about this?

**Comment by GoTcWang** [comment]
Ping this thread. 
Is it okay to both of you that we submit the fix first and refactor it later assuming release cut is tomorrow?

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (alabiaga) [reply extensions/amp-form/0.1/amp-form.js]: it does not complain about it. removed. Will fix in the next PR along with your other comments.

**Comment by GoTcWang** [comment]
Thanks Aaron and Will!

**Pull request merged**